src/escrow.rs

```diff
@@ -0,0 +1,231 @@
+        Lock {
+            id: id.to_owned(),
+            locked: Some(state),
+        }
+        .emit();
+    }
+
+    fn unlock(&mut self, id: &Self::Id, lock_handler: impl FnOnce(&Self::State) -> bool) {
+        let lock = self
+            .get_locked(id)
+            .unwrap_or_else(|| panic_str(ESCROW_NOT_LOCKED_MESSAGE));
+
+        if lock_handler(&lock) {
+            self.set_unlocked(id);
+            Lock::<_, Self::State> { id, locked: None }.emit();
+        } else {
+            panic_str(ESCROW_LOCK_HANDLER_FAILED_MESSAGE)
+        }
+    }
+
+    fn is_locked(&self, id: &Self::Id) -> bool {
+        self.get_locked(id).is_some()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Escrow;
+    use crate::escrow::EscrowInternal;
+    use near_sdk::{
+        near_bindgen, test_utils::VMContextBuilder, testing_env, AccountId, Balance, VMContext,
+        ONE_YOCTO,
+    };
+    use near_sdk_contract_tools_macros::Escrow;
+
+    const ID: u64 = 1;
+    const IS_NOT_READY: bool = false;
+
+    #[derive(Escrow)]
+    #[escrow(id = "u64", state = "bool", crate = "crate")]
+    #[near_bindgen]
+    struct Contract {
+        is_ready: bool,
+    }
+
+    #[near_bindgen]
+    impl Contract {
+        #[init]
+        pub fn new() -> Self {
+            Self { is_ready: false }
+        }
+    }
+
+    fn get_context(attached_deposit: Balance, signer: Option<AccountId>) -> VMContext {
+        VMContextBuilder::new()
+            .signer_account_id(signer.clone().unwrap_or("alice".parse().unwrap()))
+            .predecessor_account_id(signer.unwrap_or("alice".parse().unwrap()))
+            .attached_deposit(attached_deposit)
+            .is_view(false)
+            .build()
+    }
+
+    #[test]
+    fn test_can_lock() {
+        testing_env!(get_context(ONE_YOCTO, None));
+        let mut contract = Contract::new();
+
+        contract.lock(&ID, &IS_NOT_READY);
+        assert!(contract.get_locked(&ID).is_some());
+    }
+
+    #[test]
+    #[should_panic(expected = "Already locked")]
+    fn test_cannot_lock_twice() {
+        testing_env!(get_context(ONE_YOCTO, None));
+        let mut contract = Contract::new();
+
+        contract.lock(&ID, &IS_NOT_READY);
+        contract.lock(&ID, &IS_NOT_READY);
+    }
+
+    #[test]
+    fn test_can_unlock() {
+        testing_env!(get_context(ONE_YOCTO, None));
+        let mut contract = Contract::new();
+
+        let is_ready = true;
+        contract.lock(&ID, &is_ready);
+        contract.unlock(&ID, |readiness| readiness == &is_ready);
+
+        assert!(contract.get_locked(&ID).is_none());
+    }
+
+    #[test]
+    #[should_panic(expected = "Lock handler failed, not unlocking")]
+    fn test_cannot_unlock_until_ready() {
+        testing_env!(get_context(ONE_YOCTO, None));
+        let mut contract = Contract::new();
+
+        let is_ready = true;
+        contract.lock(&ID, &IS_NOT_READY);
+        contract.unlock(&ID, |readiness| readiness == &is_ready);
```

Comment:
I'm all for the elimination of magic values, but booleans are pretty straightforward, no?